### PR TITLE
Multichannel training for longitudinal studies

### PR DESCRIPTION
**Pull request by mpompolas**
This is PR to allow longitudinal studies/datasets to be trained when the Multichannel option is selected.

The file structure of these datasets includes folders and filenames with "ses-*" strings.

The derivatives are loaded in a straightforward way when they are present for all images.
However, since the multichannel training requires a single GT for multiple input images, there are a few options for associating a derivative to an image (the code follows this order):
1. The derivative of the image (normal approach)
2. The derivative of the same CONTRAST from a different session
3. The derivative from a different contrast (but also selected in the config file) from the same session
4. The derivative from a different contrast (but also selected in the config file) from a different session

The reasoning behind this is that with the previous structure of the code the images that didn't have a GT would be ignored.
However in the multi-channel training case we can still use them.

There is a display that is logged on luguru to inform the user/devs of the association of an image to a different GT.

The inspiration for accommodating these cases comes from the MS-Challenge dataset that only has a single GT for the second time point.
https://github.com/ivadomed/ms-challenge-2021

This is still in draft mode
The training/testing work
Segmenting is still WIP

Related to https://github.com/ivadomed/ivadomed/issues/788

Comments welcome.

### CODE REVIEW

**Review by jcohenadad — CHANGES_REQUESTED** [review]
it would be great to add a unit test that covers the issue and show how it resolves with the proposed changes

**Review by mariehbourget — COMMENTED** [review]
Thank you @mpompolas for looking into this!
I left a few comments, my main concern is that the code is very specific to the MS-challenge project and would probably not generalized well to other projects.

You are right that the dataframe does not include the files with missing derivatives (for MS-challenge: files from the first timepoint). I had not thought about that before. So there are like 2 main issues to resolve:
1) How to modify the filtering of the dataframe to include those files 
2) How to fill the multichannel dict. From what I understand, this is not yet addressed in this PR, and I'm not sure how this works with training.

Maybe a way to help generalization would be to add a parameter on the config to specify on "what" we want to do the multichannel (ex: contrast, session, acq, etc)?
Then compare metadata fields in the dataframe instead of relying on the filenames both for indexing and filling the multichannel dict?

**Comment by mpompolas** [comment]
thanks @jcohenadad
I'll have to create a new dataset with fake namings I guess since we don't have any publicly available longitudinal data to add for automatic download through: https://github.com/ivadomed/ivadomed/blob/master/ivadomed/scripts/download_data.py#L156

@mariehbourget you mentioned you had a few microscopy files with ses- that can help generalize the script, but I can't locate the message you sent. Can you please remind me?

File: ivadomed/loader/utils.py [hunk 1]
```diff
@@ -835,7 +838,51 @@ def get_derivatives(self, subject_fname, deriv_fnames):
...
+                        re_ordered_session_lst = session_list
+                    for session in re_ordered_session_lst:
+                        # Contrast list is reordered to first check for the same contrast-type in the other sessions
+                        if len(self.contrast_lst) > 1:
+                            re_ordered_contrast_lst = [contrast_id] + self.contrast_lst.remove(contrast_id)
+                        else:
+                            re_ordered_contrast_lst = self.contrast_lst
+                        for contrast in re_ordered_contrast_lst:
+                            new_prefix_fname = subject_fname.split('.')[0].split('_')[0] + \
+                                               "_" + session + "_" + contrast
```
[the remark this answers is not in the document]
Comment (mpompolas) [reply ivadomed/loader/utils.py:873]: Thanks for pointing this out.
I will take these into account

File: ivadomed/loader/utils.py [hunk 2]
```diff
@@ -835,7 +838,51 @@ def get_derivatives(self, subject_fname, deriv_fnames):
...
+                prefix_fname = subject_fname.split('_')[0] + "_" + contrast
+                output = [d for d in deriv_fnames if prefix_fname in d]
+                if output:
+                    break
+            if not output:
+                if "_ses-" in subject_fname:
+                    subject = subject_fname.split('_')[0]
+                    sess_id = subject_fname.split('_')[1]
+                    contrast_id = subject_fname.split('_')[-1].split(".")[0]
+                    all_subject_deriv = [d for d in deriv_fnames if subject in d]
```
[the remark this answers is not in the document]
Comment (mpompolas) [reply ivadomed/loader/utils.py:856]: Yeap, same issue as your previous point. I'll consider this here too

**Comment by mariehbourget** [comment]
> @mariehbourget you mentioned you had a few microscopy files with ses- that can help generalize the script, but I can't locate the message you sent. Can you please remind me?

Yes, they are in our data testing dataset here: https://github.com/ivadomed/data-testing/tree/master/microscopy_png/sub-rat3, only 1 subject and 2 sessions, though.

**Comment by mpompolas** [comment]
> You are right that the dataframe does not include the files with missing derivatives (for MS-challenge: files from the first timepoint). I had not thought about that before. So there are like 2 main issues to resolve:
> 
> 
>
> - How to modify the filtering of the dataframe to include those files
>
> 
>- How to fill the multichannel dict. From what I understand, this is not yet addressed in this PR, and I'm not sure how this works with training.

I just made some modifications for multichannel dict:
The association to a new derivative that I described above is inherited all the way to [fill_multichannel_dict](https://github.com/ivadomed/ivadomed/blob/8ed13e663909117dfd1d74de65b95fac0ecea031/ivadomed/loader/loader.py#L1191)

Here is an example of a variable within `fill_multichannel_dict` for an image that does not have a derivative, but inherits all the derivatives from a different session.

```
(['/home/nas/Consulting/ivado_project/Datasets/ms_challenge/LongitudinalMultipleSclerosisLesionSegmentationChallengeMiccai21/training/_curated/sub-013/ses-01/anat/sub-013_ses-01_FLAIR.nii.gz', '/home/nas/Consulting/ivado_project/Datasets/ms_challenge/LongitudinalMultipleSclerosisLesionSegmentationChallengeMiccai21/training/_curated/sub-013/ses-02/anat/sub-013_ses-02_FLAIR.nii.gz'], 
['/home/nas/Consulting/ivado_project/Datasets/ms_challenge/LongitudinalMultipleSclerosisLesionSegmentationChallengeMiccai21/training/_curated/derivatives/labels/sub-013/ses-02/anat/sub-013_ses-02_FLAIR_lesion-manual-rater1.nii.gz'], 
None, 

[{'path': '/home/nas/Consulting/ivado_project/Datasets/ms_challenge/LongitudinalMultipleSclerosisLesionSegmentationChallengeMiccai21/training/_curated/sub-013/ses-01/anat/sub-013_ses-01_FLAIR.nii.gz', 'filename': 'sub-013_ses-01_FLAIR.nii.gz', 'datatype': 'anat', 'extension': '.nii.gz', 'participant_id': 'sub-013', 'session': '01', 'subject': '013', 'suffix': 'FLAIR', 'contrast': 'FLAIR',
'input_filenames': '/home/nas/Consulting/ivado_project/Datasets/ms_challenge/LongitudinalMultipleSclerosisLesionSegmentationChallengeMiccai21/training/_curated/sub-013/ses-01/anat/sub-013_ses-01_FLAIR.nii.gz', 'gt_filenames': ['/home/nas/Consulting/ivado_project/Datasets/ms_challenge/LongitudinalMultipleSclerosisLesionSegmentationChallengeMiccai21/training/_curated/derivatives/labels/sub-013/ses-02/anat/sub-013_ses-02_FLAIR_lesion-manual-rater1.nii.gz'], 'slice_index': 255, 'coord': None}, 

{'path': '/home/nas/Consulting/ivado_project/Datasets/ms_challenge/LongitudinalMultipleSclerosisLesionSegmentationChallengeMiccai21/training/_curated/sub-013/ses-02/anat/sub-013_ses-02_FLAIR.nii.gz', 'filename': 'sub-013_ses-02_FLAIR.nii.gz', 'datatype': 'anat', 'extension': '.nii.gz', 'participant_id': 'sub-013', 'session': '02', 'subject': '013', 'suffix': 'FLAIR', 'contrast': 'FLAIR', 
'input_filenames': '/home/nas/Consulting/ivado_project/Datasets/ms_challenge/LongitudinalMultipleSclerosisLesionSegmentationChallengeMiccai21/training/_curated/sub-013/ses-02/anat/sub-013_ses-02_FLAIR.nii.gz', 'gt_filenames': ['/home/nas/Consulting/ivado_project/Datasets/ms_challenge/LongitudinalMultipleSclerosisLesionSegmentationChallengeMiccai21/training/_curated/derivatives/labels/sub-013/ses-02/anat/sub-013_ses-02_FLAIR_lesion-manual-rater1.nii.gz'], 'slice_index': 255, 'coord': None}])
```

This is still WIP.

File: ivadomed/loader/utils.py [hunk 3]
```diff
@@ -835,7 +838,51 @@ def get_derivatives(self, subject_fname, deriv_fnames):
...
+                        else:
+                            re_ordered_contrast_lst = self.contrast_lst
+                        for contrast in re_ordered_contrast_lst:
+                            new_prefix_fname = subject_fname.split('.')[0].split('_')[0] + \
+                                               "_" + session + "_" + contrast
+                            output = [d for d in deriv_fnames if new_prefix_fname in d]
+                            if output:
+                                logger.info(
+                                    "Multichannel training: No derivative found for {} "
+                                    "- Assigning derivatives from {}".format(
```
[the remark this answers is not in the document]
Comment (mpompolas) [reply ivadomed/loader/utils.py:878]: This was added with the latest commit.

**Comment by mpompolas** [comment]
Personal note:
the number of the model inputs is now dependent on the number of sessions within the dataset.
In other words, the dataset dictates these values and not the config file:
https://github.com/ivadomed/ivadomed/blob/master/ivadomed/main.py#L169

Things get a bit more complicated when a dataset has a combination of longitudinal and normal subjects that have one or multiple contrasts (even a single contrast can be located into multiple sessions, affecting the # of inputs)
Discuss strategy on how to approach this.

Add a maximum number of inputs?

**Comment by dyt811** [comment]
A quick update on this PR status. 

A minor issue was blocking the unit tests from passing, which has been fixed (dict key issue) and should be passing existing unit tests without major issues. 

HOWEVER, the BIGGER issue is 

1) **We do not have a proper unit/integration test for multi-sessions multi-contrasts currently.** 

2) Some radical attempts to adopt a small 3 subjects, 4~5 sessions, T1/T2/FLAIR/PD subset of multi-sessions-contrasts scavenged MS2015 challenge dataset (Public domain at https://smart-stats-tools.org/lesion-challenge-2015) is currently ongoing at https://github.com/MotionCorrect/data_multi-sessions-contrasts and will be migrated to IvadoMed once they are fully complete. 

We are running into issues getting the MS subjects' data loaded for some unknown reasons (despite derivatives being quite well loaded. This is going to take some quite time as from what I have seen as it is even more complicated than the MS2021 dataset. 

However, to truly properly implement multi-sessions support and user configurable session specification, a lot more remains in that direction. You can view some of the more radical extreme changes being introduced at `yd/sessions_to_multichannel` branch along with earlier attempt to adopt the dataset (since we don't have a way to auto download the 1.5GB test set yet). A lot of streamlining of that is also absolutely needed in addition to try it out on MS2021 `_curated` dataset once the access issue has been sorted out. I am still struggle to properly form the multi-sessional, multi-contrasts BIDSDict and I will need substantial more time to investigate that issue. 

Depending on the needs of the competition, I am uncertain if what Konstantinos has completed is good enough as it is to merge or we need to go to full throttle with multi-contrast/sessions support as started on the other branch.
 
Lastly, just want to be clear, pretty much ALL changes in this branch are introduced by Konstantinos. My more crazier experimental stuff are not anywhere here...

**Comment by mariehbourget** [comment]
> I am still struggle to properly form the multi-sessional, multi-contrasts BIDSDict and I will need substantial more time to investigate that issue.
> 
> Depending on the needs of the competition, I am uncertain if what Konstantinos has completed is good enough as it is to merge or we need to go to full throttle with multi-contrast/sessions support as started on the other branch.

If I recall correctly from the last meetings, the feature implemented in this branch by @mpompolas was working for the purpose of the challenge, in this https://github.com/ivadomed/ivadomed/pull/814#issuecomment-852551171, the `filename_pairs` seems correct.

Our concern with merging was that the feature was specific to the challenge and not easily generalizable to other situation. If that is the case, would it be possible to do the training for the challenge on this branch without merging?

**Comment by mpompolas** [comment]
I managed to get it working for the challenge with a hardcoded entry for the number of inputs on:
https://github.com/ivadomed/ivadomed/blob/master/ivadomed/main.py#L169
model_params["in_channel"] = 2

@dyt811 is working on generalizing on other datasets and also introducing new variables within the `config.json` file for multichannel filtering selections

**Comment by dyt811** [comment]
@mariehbourget was SPOT on for the issue. A new updated dataset with PROPER capitalization is now up there. @cakester is wrapping up the loader unit tests and I can start testing out the potential new experimental flag related to session specification from user and single source of ground truth implementation to help generalize better.

**Comment by dyt811** [comment]
More clear explanation will be given at the upcoming presentation. This is a technical brief on the direction the upcoming PR will head towards. 

0) Unit tests for loading multi-sessional-multi contrasts data are passing BUT many edge cases are not even remotely being covered because, assumptions below are key potential contentions and will be extensively discussed in the upcoming meetings. It has been proposed/added in the radical branch `yd/sessions_to_multichannels`. 

1) A session based filter key has been proposed/added. This enable restricting and targeting specific session x modality to ensure a consistent input for the model input channel for Main.py (not yet implemented). 

2) All modality data must be present for that subject to be included. 

2) A ground truth loader configure key has been proposed/added in the same branch to allow cross-session, cross-modality ground truth specification (i.e. one ground truth to rule them all PER subject). Key assumption going forward will be all data must be aligned to each other BEFORE IvadoMed. This way, it drastically reduce the complexity needed to do ground truth match across modality, across sessions. 

3) Once those edge cases are covered, the only remaining major issues is to ensure model_params["in_channel"] is properly updated. At that point, we can tentatively say that we support multi-session multi-contrasts data with huge caveat on how the data must be pre-aligned across sessions AND modalities before trying ivadomed and config.json must include details about the new `target_ground_truth` and `target_sessions` key for the model to know: 1) what ground truth to consider, 2) which sessions data to ensure are present.

**Comment by dyt811** [comment]
We have tentatively achieved what we set up in terms of subject/session/contrast inclusion/exclusion approach set out and built a full suite of unit tests leverage pytest-cases around it and they are passing in the other branch.  
image

It was quite radical/experimental and may not be exactly what people have in mind when it comes to supporting multi-contrasts/sessions. However, it is a point to start and require much further refinement and improvement in the upcoming days before it can be ready to integrate with the rest of the main ivadomed code due to amount of changes it introduce (i.e. a new mandatory config keywords).  

The presentation summarizing the approach is here: https://docs.google.com/presentation/d/1yJJTsciiRCPSw9bRUUPpEmqle6jIO2qcfiuRALYAL4s/edit?usp=sharing

**Comment by mariehbourget** [comment]
Thank you @dyt811 and @cakester for the report and update, it helps a lot to understand the different scenarios. I have a few questions/comments:
1)	I don’t have any experience with multichannel training so I’m not sure about the behavior on slides 5, 11 and 12 i.e. everytime we keep a GT from a modality or session that is not specified in the config file. I think we should confirm/discuss with the ivadomed team.

2)	In the presentation, there is always only 1 ground-truth available for each multichannel input. What happens when several GT are available, is the “best” match in terms of modality/sessions is kept?

3)	I’m not sure I understand the purpose of `target_ground_truth` and the difference with `target_suffix`. When there are several `target_suffix`, it means that we train a multi-class model, so we need 1 GT per class (each `target_suffix` represents a class). So unless I’m mistaken, multi-channel and multi-class are two separate things and if we filter with a single “target_ground_truth”, we won’t be able to do multi-class training.

4)	If I understand correctly, the new features are implemented in another branch (`yd/sessions_to_multichannel`), which also includes a major refactoring of the loader. If that is the case, maybe we should open a separate draft PR to be able to visualize the changes and discuss further your approach.

**Comment by dyt811** [comment]
So to summarize the current direction this PR will take. Per last week's discussion here are the clarifications on the questions. 
 
> 1. I don’t have any experience with multichannel training so I’m not sure about the behavior on slides 5, 11 and 12 i.e. everytime we keep a GT from a modality or session that is not specified in the config file. I think we should confirm/discuss with the ivadomed team.

The team concurred on some of the basic key assumptions: 
1) We can safely assume all Session and Modalities are properly aligned. 
2) We will not use a new key but instead we will ensure that at least ONE of EACH of the "target_suffix" key is present at the subject level (across Sessions and Modalities). 
3) We require user to specify the new specific target_sessions key to use. We require user to specify the modalities required. The mandatory requirements are used to help ensure the right model input channels. If missing one modality or session, we discard subject.

> 2. In the presentation, there is always only 1 ground-truth available for each multichannel input. What happens when several GT are available (ex: GT for sessions 1 and 2 are available), is the “best” match in terms of modality/sessions with the input is kept?

So far, we are defaulting to the very first GT if more than one is specified. 

> 3. I’m not sure I understand the purpose of `target_ground_truth` and the difference with `target_suffix`. When there are several `target_suffix`, it means that we train a multi-class model, so we need 1 GT per class (each `target_suffix` represents a class). So unless I’m mistaken, multi-channel and multi-class are two separate things and if we filter with a single “target_ground_truth”, we won’t be able to do multi-class training.

The target ground truth key has been removed in favour of repurposing the target_suffix instead. 

> 4. If I understand correctly, the new features are implemented in another branch (`yd/sessions_to_multichannel`), which also includes a major refactoring of the loader. If that is the case, maybe we should open a separate draft PR to be able to visualize the changes and discuss further your approach.

Yes. As soon as an end to end training is completed successfully, we will breakdown the new branch into bite size chunk PRs to the main branch to ease the adaption. So far, we have not been able to even achieve end to end training EVEN with the feature branch so as soon as we hit that mile stone, we will start migrating it towards the main branch.

**Comment by dyt811** [comment]
This PR is now being closed as we are **reforging** it in a new PR **post** the major `loader.py` class refactoring. 

Rest assured, **ALL changes** introduced here as of 2021-12-12T115559EST has been manually checked and migrated over to the new PR branch as of commit `df1e319dd4700a50a50a8320439ce030320b5730` made on `yd/multi_session_support_reforged` branch which will subsume existing work from `kn/sessions_to_multichannel` as well as the more radical exploratory branch formed earlier `yd/sessions_to_multichannel` as part of [Project 5](https://github.com/ivadomed/ivadomed/projects/5) In progress items.

**Comment by dyt811** [comment]
Do not delete this branch until the new PR is fully merged and ready.
